File: library/src/main/java/com/flipkart/okhttpstats/NetworkInterceptor.java
/*
 * The MIT License
 *
 * Copyright 2016 Flipkart Internet Pvt. Ltd.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this
 * software and associated documentation files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom
 * the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.flipkart.okhttpstats;

import com.flipkart.okhttpstats.interpreter.NetworkInterpreter;
import com.flipkart.okhttpstats.toolbox.Utils;

import java.io.IOException;
import java.util.concurrent.atomic.AtomicInteger;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

/**
 * This class is the entry point for this library.
 * A OKHttp {@link Interceptor} which intercepts all requests & responses and passes it on to the {@link NetworkInterpreter} along with {@link TimeInfo}
 * All network related exceptions are also bubbled to the same interpreter.
 */
public final class NetworkInterceptor implements Interceptor {

    private final NetworkInterpreter mInterpreter;
    private final AtomicInteger mNextRequestId = new AtomicInteger(1);
    private final boolean mEnabled;

    NetworkInterceptor(Builder builder) {
        mEnabled = builder.mEnabled;
        if (builder.mInterpreter == null) {
            throw new IllegalStateException("NetworkInterpreter cannot be null");
        }
        mInterpreter = builder.mInterpreter;
        Utils.isLoggingEnabled = builder.mIsLoggingEnabled;
    }

    /**
     * Method to intercept all the network calls.
     *
     * @param chain {@link Interceptor.Chain}
     * @return {@link Response}
     * @throws IOException
     */
    @Override
    public Response intercept(Chain chain) throws IOException {
        final int requestId = mNextRequestId.getAndIncrement();
        Request request = chain.request();
        TimeInfo timeInfo = new TimeInfo();
        Response response;
        try {
            //note the time taken for the response
            timeInfo.mStartTime = System.currentTimeMillis();
            response = chain.proceed(request);
            timeInfo.mEndTime = System.currentTimeMillis();
        } catch (final IOException e) {
            if (mEnabled) {
                //notify event reporter in case there is any exception while proceeding request
                mInterpreter.interpretError(requestId, timeInfo, request, e);
            }
            throw e;
        }

        if (mEnabled) {
            response = mInterpreter.interpretResponseStream(requestId, timeInfo, request, response);
        }

        return response;
    }

    public static class TimeInfo {
        public long mStartTime;
        public long mEndTime;
    }

    /**
     * Builder Pattern for {@link NetworkInterceptor}
     */
    public static class Builder {
        boolean mEnabled = true;
        boolean mIsLoggingEnabled = false;
        NetworkInterpreter mInterpreter;

        /**
         * To enable/disable the calls to {@link NetworkInterpreter}
         * If disabled, the interceptor continues to operate without reporting to the {@link NetworkInterpreter}
         *
         * @param enabled boolean
         * @return {@link Builder}
         */
        public Builder setEnabled(boolean enabled) {
            this.mEnabled = enabled;
            return this;
        }

        /**
         * Can leave it null for the default implementation
         *
         * @param interpreter {@link NetworkInterpreter}
         * @return {@link Builder}
         */
        public Builder setNetworkInterpreter(NetworkInterpreter interpreter) {
            this.mInterpreter = interpreter;
            return this;
        }

        /**
         * To enable/disable logging. By default logging is disabled.
         *
         * @param isLoggingEnabled : boolean
         * @return {@link Builder}
         */
        public Builder setLoggingEnabled(boolean isLoggingEnabled) {
            this.mIsLoggingEnabled = isLoggingEnabled;
            return this;
        }

        public NetworkInterceptor build() {
            return new NetworkInterceptor(this);
        }
    }
}

File: library-kotlin/src/main/java/com/flipkart/okhttpstats/kotlin/NetworkInterceptor.kt
package com.flipkart.okhttpstats.kotlin

import com.flipkart.okhttpstats.kotlin.interpreter.NetworkInterpreter
import com.flipkart.okhttpstats.kotlin.toolbox.Utils
import okhttp3.Interceptor
import okhttp3.Response
import java.io.IOException
import java.util.concurrent.atomic.AtomicInteger

class NetworkInterceptor(builder: Builder) : Interceptor {

    private var interpreter: NetworkInterpreter
    private val nextRequestId = AtomicInteger(1)
    private var enabled = true

    init {
        enabled = builder.enabled
        if (builder.interpreter == null) {
            throw IllegalStateException("NetworkInterpreter cannot be null")
        }
        interpreter = builder.interpreter!!
        Utils.isLoggingEnabled = builder.isLoggingEnabled
    }

    override fun intercept(chain: Interceptor.Chain): Response {
        val requestId = nextRequestId.getAndIncrement()
        val request = chain.request()
        val timeInfo = TimeInfo()
        var response: Response
        try {
            //note the time taken for the response
            timeInfo.startTime = System.currentTimeMillis()
            response = chain.proceed(request)
            timeInfo.endTime = System.currentTimeMillis()
        } catch (e: IOException) {
            if (enabled) {
                //notify event reporter in case there is any exception while proceeding request
                interpreter.interpretError(requestId, timeInfo, request, e)
            }
            throw e
        }

        if (enabled) {
            response = interpreter.interpretResponseStream(requestId, timeInfo, request, response)
        }

        return response
    }

    class TimeInfo {
        var startTime: Long = 0
        var endTime: Long = 0
    }

    /**
     * Builder Pattern for [NetworkInterceptor]
     */
    class Builder {
        internal var enabled = true
        internal var isLoggingEnabled = false
        internal var interpreter: NetworkInterpreter? = null

        /**
         * To enable/disable the calls to [NetworkInterpreter]
         * If disabled, the interceptor continues to operate without reporting to the [NetworkInterpreter]

         * @param enabled boolean
         * *
         * @return [Builder]
         */
        fun setEnabled(enabled: Boolean): Builder {
            this.enabled = enabled
            return this
        }

        /**
         * Can leave it null for the default implementation

         * @param interpreter [NetworkInterpreter]
         * *
         * @return [Builder]
         */
        fun setNetworkInterpreter(interpreter: NetworkInterpreter): Builder {
            this.interpreter = interpreter
            return this
        }

        /**
         * To enable/disable logging. By default logging is disabled.

         * @param isLoggingEnabled : boolean
         * *
         * @return [Builder]
         */
        fun setLoggingEnabled(isLoggingEnabled: Boolean): Builder {
            this.isLoggingEnabled = isLoggingEnabled
            return this
        }

        fun build(): NetworkInterceptor {
            return NetworkInterceptor(this)
        }
    }
}